### PR TITLE
(PXP-6815): update MDS object after file upload.

### PR DESCRIPTION
**Pull request by johnfrancismccann**
Extend indexs3client job to update the `_bucket`, `_filename`, `_file_extension`, and `_upload_status` keys in corresponding Metadata Service object after a file has been uploaded to the bucket.

[Jira ticket](https://ctds-planx.atlassian.net/browse/PXP-6815)

Metadata Service config must be nested in the `CONFIG_FILE` environment variable:
```
{
        "url": "http://indexd-service/index",
	"username": "mr happy cat",
	"password": "whiskers",
	"metadataService": {
		"url": "http://revproxy-service/mds",
		"username": "mr friendly cat",
		"password": "paws"
	}
}
```

~PLEASE DO NOT MERGE YET. [kube-setup-ssjdispatcher PR to set up MDS creds](https://github.com/uc-cdis/cloud-automation/pull/1428) must be merged first.~ Additionally, this indexs3client PR should go out in the same monthly release as the [SSJDispatcher ticket/PR to require MDS creds on startup](https://ctds-planx.atlassian.net/browse/PXP-6977).

### New Features
- Update the `_bucket`, `_filename`, `_file_extension`, and `_upload_status`
 keys in corresponding Metadata Service object after a file has been uploaded to the bucket.

### Breaking Changes
- Going forward, indexs3client now requires the Metadata Service to be deployed. Metadata Service creds are now required in the `CONFIG_FILE` environment variable in order for this job to run. Without both Indexd and MDS creds configured, this job will immediately exit and update neither Indexd or MDS.

### Improvements
- Determine if Indexd record already has hashes and size populated before downloading S3 object and calculating hashes. If so, don’t bother downloading/recalculating.
- Define `ConfigInfo struct` so that the config (url and creds) for new services can be easily unmarshalled from the `CONFIG_FILE` environment variable
- Make use of `retryablehttp` package to abstract http request retry logic and only retry on 500-range errors

### Dependency updates
- Add `github.com/hashicorp/go-retryablehttp v0.6.7` to `go.mod`

### Deployment changes
- `metadata-service` must already be present or added to the `"versions": {` section of the `cdis-manifest`. After `metadata-service` has been confirmed to be in the `cdis-manifest`, the now-required Metadata Service creds for SSJDispatcher/indexs3client must be configured by running `gen3 roll all` with `JENKINS_HOME` **not** set. `gen3 roll all` calls `gen3 kube-setup-ssjdispatcher`, which now [automatically configures](https://github.com/uc-cdis/cloud-automation/pull/1428) the SSJDispatcher secret with the appropriate Metadata Service creds. After running `gen3 roll all`, SSJDispatcher should be configured such that SSJDispatcher can automatically pass the Metadata Service creds to the indexs3client job instance (i.e. after running `gen3 roll all`, no further action should be required by DevOps).

### CODE REVIEW

**Review by giangbui — APPROVED** [review]
LGTM!

File: .secrets.baseline [hunk 1]
```diff
@@ -58,68 +58,241 @@
     }
   ],
   "results": {
-    "Gopkg.lock": [
+    "go.sum": [
```
Comment (paulineribeyre) [inline .secrets.baseline:61]: you can exclude `go.sum` from the scanning

File: handlers/metadata.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"bytes"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+// Checks for presence of MDS config, calls UpdateMetadataObject, and logs info
+func updateMetadataObjectWrapper(uuid string, configInfo *ConfigInfo, body string) {
```
[the remark this answers is not in the document]
Comment (johnfrancismccann) [reply handlers/metadata.go:12]: Yeah that was on purpose. The idea is that this wrapper does all the error handling and logging for the caller. I had been calling `UpdateMetadataObject` from `handlers/handler.go` in three places and just wanted to DRY.

File: handlers/handler.go [hunk 3]
```diff
@@ -63,6 +84,21 @@ func IndexS3Object(s3objectURL string) {
...
+	if err != nil {
+		log.Panicf("Can not get record %s from Indexd. Error message %s", uuid, err)
+	} else if rev == "" {
+		log.Printf("Indexd record with guid %s already has size and hashes", uuid)
+		updateMetadataObjectWrapper(uuid, configInfo, mdsUploadedBody)
+		return
+	}
+	log.Printf("Got rev %s from Indexd for record %s", rev, uuid)
+
+	updateMetadataObjectWrapper(uuid, configInfo, `{"_upload_status": "indexs3client job calculating hashes and size"}`)
```
[the remark this answers is not in the document]
Comment (johnfrancismccann) [reply handlers/handler.go:101]: Yeah just one thing to keep in mind is what `_upload_status` should be in the case that this job fails to download and calculate hashes for the file object (e.g. maybe the file is too big and can't be downloaded)

[the remark this answers is not in the document]
Comment (paulineribeyre) [reply handlers/handler.go:101]: yeah, that's not detailed in the design doc, i'll let you and Alex make that call

File: handlers/handler.go [hunk 4]
```diff
@@ -63,6 +84,21 @@ func IndexS3Object(s3objectURL string) {
 	} else {
 		uuid = strings.Join(split_key[:len(split_key)-1], "/")
 	}
+	filename := split_key[len(split_key)-1]
+
+	log.Printf("Attempting to get rev for record %s in Indexd", uuid)
+	rev, err := GetIndexdRecordRev(uuid, configInfo.Indexd.URL)
+	mdsUploadedBody := fmt.Sprintf(`{"_upload_status": "uploaded", "_filename": "%s"}`, filename)
```
[the remark this answers is not in the document]
Comment (johnfrancismccann) [reply handlers/handler.go:91]: I talked with @Avantol13 about this. We thought `_filename` could be preferable. I just updated the design doc to reflect this.

**Review by giangbui — APPROVED** [review]

**Pull request merged**